core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -215,6 +215,12 @@ public void initialize(String name, Map<String, String> unresolved) {
     this.paths = ResourcePaths.forCatalogProperties(mergedProps);
 
     String token = mergedProps.get(OAuth2Properties.TOKEN);
+    // re-resolve these variables in case they were overridden by the config endpoint
+    credential = mergedProps.get(OAuth2Properties.CREDENTIAL);
```

Comment:
It is weird to me that the client gets the credential from server's config endpoint, which doesn't need authentication. Does that mean any client can visit the REST catalog? I think we still need a client to provide its own credential.

Reply:
I'd argue that the client can already get a `token` from the server's config endpoint, so why not a `credential`? :-) 

Taking a step back, I'd like this PR to focus on what the _client_ is supposed to do with configuration received from the server, not what the _server_ is supposed to send. For that, there is an ongoing discussion [here](https://docs.google.com/document/d/1JUtFpdEoa6IAKt1EzJi_re0PUbh56XnfUtRe5WAfl0s/edit?usp=sharing) where @jbonofre and others suggest many changes, e.g. the introduction of a "handshake" or "pre-flight" endpoint to facilitate client auto-configuration, including auth properties. But server-side changes will take a lot more time.

And back to the client side of things: I think it makes sense for the client to apply _all_ configuration received from the server. It's then up to the server to decide which options make sense to send, and which don't.

One important property I'd like to see properly handled is `oauth2-server-uri`. I'd also like to see the map `optionalOAuthParams` correctly handled, since it may contain more important config properties in the future.

Reply:
I think line 217 is to get the token from client config other than from the server config response. `mergedProps` mixes properties from both client config and server config response.

Also I believe a client should only get tokens either from users' config or from token server URI response. Config endpoint shouldn't be used for passing tokens. Otherwise, every client may use the same token/credentials, which doesn't make sense to me.

Reply:
I would second @flyrain's original comment that the client should not be exchanging credentials via this path.  The client credentials flow is for auth "previously arranged with the authorization server".  The token should be used if returned by the server, but credentials should client controlled.

Reply:
The `/v1/config` endpoint - titled as "List **all** catalog configuration settings" - seems to [require authentication](https://github.com/apache/iceberg/blob/ea916c1700ef37377d3e8bba1197b3fe32d1a248/open-api/rest-catalog-open-api.yaml#L121-L122). Catalog configuration settings can already contain sensitive information, or information that you do not want everybody to know (warehouse location for example) - so it makes a lot of sense to require authentication for that endpoint.

Reply:
Also, the config endpoint seems to already return sensitive information - in line 217 above (`String token = mergedProps.get(OAuth2Properties.TOKEN);`)

Reply:
> seems to [require authentication](https://github.com/apache/iceberg/blob/ea916c1700ef37377d3e8bba1197b3fe32d1a248/open-api/rest-catalog-open-api.yaml#L121-L122).

There is no requirement that use of the REST protocol requires auth.  For people operating their own internal REST implementations, they can choose to support auth or not.

> the config endpoint seems to already return sensitive information

The config endpoint may return sensitive information or not, but I won't recommend returning sensitive information if the communication channel is not secure.

Reply:
The resource (here: Iceberg REST endpoint) _is_ (but should have really never been) the de-facto authentication server, because of the existence of the existence of the `/v1/oauth/tokens` endpoint. So the argument that the config-resource should not return credentials feels odd.

Reply:
I don't think it's actually correct to send back a `credential` to the client via the `config` endpoint. The client credential flow follows the OAuth2 Spec as defined [here](https://datatracker.ietf.org/doc/html/rfc6749#autoid-49) (there's also a shorter example for this [here](https://www.oauth.com/oauth2-servers/access-tokens/client-credentials/)) and so sending back a client-controlled `credential` to the client doesn't feel right

Reply:
And yet, that is exactly what is being done [here](https://github.com/apache/iceberg/blob/7caedf85954c78fb3212414b924ff29302990b2a/core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java#L945), but from the tables endpoint.